timer_spec: no timeout with "doesn't mess up the cmdline"

blueyed opened the pull request:
It was increased in dd21cd2a4 to avoid flakiness, but takes 1s then.

Not specifying a timeout argument makes it pass faster.

This is due to `1000` being used as minimal timeout when a timeout is passed: https://github.com/blueyed/neovim/blob/3518bb328b7b5e8394626764f03cb400b6068a8e/test/functional/ui/screen.lua#L495

Not really sure if that's a bug by itself, or just how it is used here.

/cc @bfredl

blueyed committed:
timer_spec: no timeout with "doesn't mess up the cmdline"

It was increased in dd21cd2a4 to avoid flakiness, but takes 1s then.

Not specifying a timeout argument makes it pass faster.

blueyed committed:
rebuild

bfredl commented:
Removing the minimal timeout makes the test pass faster by not fully testing the relevant condition. The reasonable minimum value is the one I specified initially, 200ms. Then we need to deal with that OS X travis is generally unreliable for timing based tests. I would have thought `default_timeout_factor` will cover that, but it seems it will not apply to `expect{timeout=...}` arg, maybe we think the specific value has a more precise meaning then.

Perhaps there could be a `timeout_scale` flag to `expect` to apply the `default_timeout_factor` even there. Or maybe it should be done automatically for every case, at least on travis OS X.

blueyed committed:
squash! timer_spec: no timeout with "doesn't mess up the cmdline"

Wait for timer to have echoed, then compare the screen.

Test failure triggered via:

```diff
diff --git i/src/nvim/ex_getln.c w/src/nvim/ex_getln.c
index 17292dfde..d56dea1c0 100644
--- i/src/nvim/ex_getln.c
+++ w/src/nvim/ex_getln.c
@@ -546,9 +546,9 @@ static int command_line_execute(VimState *state, int key)
       do_cmdline(NULL, getcmdkeycmd, NULL, DOCMD_NOWAIT);
     }

-    if (!cmdline_was_last_drawn) {
-      redrawcmdline();
-    }
+    // if (!cmdline_was_last_drawn) {
+    //   redrawcmdline();
+    // }
     return 1;
   }

```

blueyed commented:
Hmm, the doc for intermediate says:

> intermediate:Test that the final state is the same as the previous expect,
> but expect an intermediate state that is different. If possible
> it is better to use an explicit screen:expect(...) for this
> intermediate state.

I understand that this tests for the intermediate states inbetween.
This seems to be good in general - although it is not really tested what intermediate states there are.

I've pushed a follow up commit that should harden the test in this regard, by checking that the timer is finished before expecting the non-messed up screen.

Additionally this could be added, but is not really necessary I guess:
```diff
diff --git i/test/functional/eval/timer_spec.lua w/test/functional/eval/timer_spec.lua
index e78303c96..e28b6035e 100644
--- i/test/functional/eval/timer_spec.lua
+++ w/test/functional/eval/timer_spec.lua
@@ -234,6 +234,16 @@ describe('timers', function()
       {0:~                                       }|
       :good^                                   |
     ]]}
+
+    feed("\027")
+    screen:expect{grid=[[
+      ^                                        |
+      {0:~                                       }|
+      {0:~                                       }|
+      {0:~                                       }|
+      {0:~                                       }|
+                                              |
+    ]]}
   end)

 end)
```

justinmk commented on test/functional/eval/timer_spec.lua:
```diff
@@ -220,16 +220,20 @@ describe('timers', function()
-    ]], intermediate=true, timeout=1000}
-
-    eq(1, eval('g:val'))
+    ]]}
```
Looks the same as the previous screen test. Will it warn about "unchanged" now?

blueyed replied:
Oh, true:

> warning: Screen test succeeded immediately. …

Keeping `intermediate=true` fixes it.

We might keep the initial timeout then now also I assume (to ensure it is not changed for 200ms still), and the extra check added (waiting for the timer to have finished) should then fix the initial flakiness on osx.

justinmk replied:
> Keeping `intermediate=true` fixes it.

I would think `unchanged=true` should be used. Perhaps `intermediate=true` works by accident...

blueyed replied:
`unchanged=true` fails with intermediate states.

Using `screen:redraw_debug()` shows:
```
screen:expect{grid=[[
                                          |
  {0:~                                       }|
  {0:~                                       }|
  {0:~                                       }|
  {0:~                                       }|
  :good^                                   |
]]}

{ "grid_line", { 1, 5, 0, { { "e", 0 }, { "v" }, { "i" }, { "l" } } }, { 1, 5, 4, { { " ", 0, 1 } } } }
{ "grid_cursor_goto", { 1, 5, 4 } }
{ "flush", {} }

screen:expect{grid=[[
                                          |
  {0:~                                       }|
  {0:~                                       }|
  {0:~                                       }|
  {0:~                                       }|
  evil^                                    |
]]}

{ "grid_line", { 1, 5, 0, { { ":", 0 } } }, { 1, 5, 1, { { "g", 0 }, { "o", 0, 2 }, { "d" } } } }
{ "grid_cursor_goto", { 1, 5, 5 } }
{ "flush", {} }

screen:expect{grid=[[
                                          |
  {0:~                                       }|
  {0:~                                       }|
  {0:~                                       }|
  {0:~                                       }|
  :good^                                   |
]]}
```

i.e. "evil" shows up, but get overwritten then apparently.  Asserting this is flaky though (i.e. not seen all the time)

@bfredl 
Is that how it should work?  Or should "evil" never be echoed in the first place really?

bfredl replied:
@blueyed If "evil" was not echoed then this test would have tested nothing at all. (though an optimization to suppress it wouldn't be wrong by itself, but then we would need to change the test to use something else that is not). The purpose is to check whether the command line is restored (by nvim itself) after being "messed up" by a timer.  Using `intermediate=true` is not somehow an "accident", it is used because it is exactly what the test is supposed to check.

blueyed committed:
keep intermediate=true

bfredl commented:
> I understand that this tests for the intermediate states inbetween.
This seems to be good in general - although it is not really tested what intermediate states there are.

@blueyed No, the point is explicitly not to "test for" the states in between, then we would just have written them out already. The purpose is to check that the final state is consistent with the initial one, but to be able to do that efficiently the test runner needs to know whether the particular situation has intermediate states or not (because the most efficient way to wait for the final state is different then).

bfredl commented on test/functional/eval/timer_spec.lua:
```diff
@@ -220,16 +220,20 @@ describe('timers', function()
     ]])
 
+    -- Wait for timer to have finished / echoed.
+    retry(nil, nil, function()
```
Adding extra RPC requests causes extra redraws. This negates the purpose of the test: to check whether nvim redraws the command line after the timer redraw completely by itself.

bfredl commented:
Was my suggestion of just scaling the timeouts with default_timeout_factor tested? This would make the test more forgiving on slow CI, while keep being "sharp" on fast systems, without needlessly changing the meaning of the test.

blueyed committed:
shorter base timer

blueyed commented:
> Was my suggestion of just scaling the timeouts with default_timeout_factor tested?

No, since it's not really trivial when I've looked at it back then IIRC (all the options to Screen:expect are quite tangled already).

> The reasonable minimum value is the one I specified initially, 200ms

I assume that is based on the timer taking 100ms?
If we shorten that to 10ms, would 20ms suffice?  (or 100ms to be really sure)

blueyed committed:
fixup! shorter base timer

blueyed committed:
back out

bfredl commented:
> No, since it's not really trivial when I've looked at it back then IIRC (all the options to Screen:expect are quite tangled already).

I can try to untangle it. In the meanwhile `load_adjust(200)` as in the latest commit probably works.

jamessan commented:
FWIW, this is semi-regularly failing on some of the slower Debian builds (e.g., [mipsel](https://buildd.debian.org/status/fetch.php?pkg=neovim&arch=mipsel&ver=0.4.3-2&stamp=1573832947&raw=0)).

justinmk commented:
We can bump it up to 1000 again.

blueyed commented:
For reference:
```
[32m[2m[ RUN      ][0m[0m timers doesn't mess up the cmdline: 
warning: Screen changes were received after the expected state. This indicates
indeterminism in the test. Try adding screen:expect(...) (or wait()) between
asynchronous (feed(), nvim_input()) and synchronous API calls.
  - Use screen:redraw_debug() to investigate; it may find relevant intermediate
    states that should be added to the test to make it more robust.
  - If the purpose of the test is to assert state after some user input sent
    with feed(), adding screen:expect() before the feed() will help to ensure
    the input is sent when Nvim is in a predictable state. This is preferable
    to wait(), for being closer to real user interaction.
  - wait() can trigger redraws and consequently generate more indeterminism.
    Try removing wait().
      
stack traceback:
	test/functional/ui/screen.lua:589: in function '_wait'
	test/functional/ui/screen.lua:370: in function 'expect'
	test/functional/eval/timer_spec.lua:232: in function <test/functional/eval/timer_spec.lua:218>

[1m[31mERR[0m[0m
test/functional/ui/screen.lua:595: expected intermediate screen state before final screen state

stack traceback:
	test/functional/ui/screen.lua:595: in function '_wait'
	test/functional/ui/screen.lua:370: in function 'expect'
	test/functional/eval/timer_spec.lua:241: in function <test/functional/eval/timer_spec.lua:218>
```

Code: https://github.com/neovim/neovim/blob/924fa143ccd4e144a95a0dd2630e609f896f0c59/test/functional/eval/timer_spec.lua#L230-L248

So the problem there does not appear to be the timeout that was changed here, but that the intermediate step is seen before already?

blueyed commented:
The code above is the release-0.4 branch.  Compared to 0.3 there's also an additional `redraw` added in https://github.com/neovim/neovim/commit/e04b9e7c78bdcf40cce5b5533ce8d7695691dfd6#diff-bc7e8dd52480925fa2d684b362fbad1bR225 that might cause/trigger this.
